activerecord/lib/active_record/connection_adapters/sqlite3_adapter.rb

```diff
@@ -365,7 +365,7 @@ def table_structure(table_name)
         # See: https://www.sqlite.org/lang_altertable.html
         # SQLite has an additional restriction on the ALTER TABLE statement
         def invalid_alter_table_type?(type, options)
-          type.to_sym == :primary_key || options[:primary_key]
+          type.to_sym == :primary_key || options[:primary_key] || !options.fetch(:null, true)
```

Comment:
https://www.sqlite.org/lang_altertable.html

> If a NOT NULL constraint is specified, then the column must have a default value other than NULL.

```suggestion
          type.to_sym == :primary_key || options[:primary_key] ||
            options[:null] == false && options[:default].nil?
```

Reply:
Good catch. Thanks!